### PR TITLE
Add support for non-pointer nullable types

### PR DESCRIPTION
**Pull request by lukap3**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -357,6 +357,16 @@ The schema of the type will look like the following instead of describing all th
    "format": "uuid"
 }
 ```
+
+If you want to override the `nullable` property of a type, you can implement the `Nullable` interface for this type.
```
Comment (wI2L) [inline README.md:361]: A link to the code definition of the interface wrapping the `Nullable` inline-code would be nice.

File: README.md [hunk 2]
```diff
@@ -357,6 +357,16 @@ The schema of the type will look like the following instead of describing all th
...
+
+For example, if `NullString` is not referenced by a pointer in your model but you still want it to be "nullable"
```
Comment (wI2L) [inline README.md:363]:
```suggestion
For example, if `NullString` is not referenced by a pointer in your model but you still want it to be "nullable":
```

File: README.md [hunk 3]
```diff
@@ -357,6 +357,16 @@ The schema of the type will look like the following instead of describing all th
...
+```go
```
Comment (wI2L) [inline README.md:364]: Add a new line before code-block.

File: README.md [hunk 4]
```diff
@@ -357,6 +357,16 @@ The schema of the type will look like the following instead of describing all th
...
+type NullSring struct { ... }
+
+func (*NullString) Nullable() bool { return true }
```
Comment (wI2L) [inline README.md:367]:
```suggestion
func (NullString) Nullable() bool { return true }
```

Doesn't need a pointer receiver in the example.

File: openapi/generator.go [hunk 5]
```diff
@@ -877,6 +877,13 @@ func (g *Generator) newSchemaFromType(t reflect.Type) *SchemaOrRef {
 		nullable = true
 	}
 	dt := g.datatype(t)
+
+	// Check for "Nullable"
```
Comment (wI2L) [inline openapi/generator.go:881]: Redundant/not useful comment.

Also, the implementation should first use `reflect.Type.Implements()` to check if the type implements the interface before creating a new value with `reflect.New` to call the `Nullable` method.

You can know the type of an interface like so:
```
tofNullable = reflect.TypeOf((*Nullable)(nil)).Elem()
```

ex: https://github.com/wI2L/fizz/blob/master/openapi/types.go#L15

which is then used like so:
```
t.Implements(tofNullable)
```

https://github.com/wI2L/fizz/blob/master/openapi/types.go#L123

File: openapi/types.go [hunk 6]
```diff
@@ -48,6 +48,12 @@ type Exampler interface {
 	ParseExample(v string) (interface{}, error)
 }
 
+// Nullable is the interface implemented by the types
+// that can be nullable
```
Comment (wI2L) [inline openapi/types.go:52]:
```suggestion
// that can be nullable.
```

**Review by wI2L — CHANGES_REQUESTED** [review]
Few changes required.
Also, please add a test with a custom type that implement the `Nullable` method: non-pointer type that can be nulled with a struct field (`NullString`).

File: README.md (hunk 1 above)
Comment (lukap3) [reply README.md:361]: Changed it a bit and added a link to https://pkg.go.dev/database/sql#NullString which hopefully makes it a bit more clear how/why this is useful 👍

File: openapi/generator.go (hunk 5 above)
Comment (lukap3) [reply openapi/generator.go]: Makes sense! Thanks 👍

File: openapi/generator.go [hunk 7]
```diff
@@ -1264,6 +1265,11 @@ func parseExampleValue(t reflect.Type, value string) (interface{}, error) {
 		return i.ParseExample(value)
 	}
 
+	if t.Name() == "NullString" {
```
Comment (wI2L) [inline openapi/generator.go:1268]: Remove this block.

File: openapi/generator.go [hunk 8]
```diff
@@ -878,10 +878,11 @@ func (g *Generator) newSchemaFromType(t reflect.Type) *SchemaOrRef {
...
 	dt := g.datatype(t)
 
-	// Check for "Nullable"
-	i, ok := reflect.New(t).Interface().(Nullable)
-	if ok {
-		nullable = i.Nullable()
+	if t.Implements(tofNullable) {
+		i, ok := reflect.New(t).Interface().(Nullable)
+		if ok {
+			nullable = i.Nullable()
```
Comment (wI2L) [inline openapi/generator.go:884]: This overrides the assignation made line 877, if the type is a pointer.
This should be taken into consideration in priority, even if the interface's method return false next.
This should be covered by a test with a pointer type to a `Nullable` interface to ensure that pointers are always marked as nullable in the spec.

**Review by wI2L — CHANGES_REQUESTED** [review]
Still missing tests to cover the various cases:
- Non-pointer to `Nullable` interface type: return true/false
- Pointer to `Nullable` interface type: return true/false

File: openapi/generator_test.go [hunk 9]
```diff
@@ -43,13 +45,15 @@ type (
 		D []*Y
 		E [3]*X
 		F *X
-		G *Y
+		G Y
```
Comment (wI2L) [inline openapi/generator_test.go:48]: Why did you change that field ?

File: openapi/generator_test.go [hunk 10]
```diff
@@ -75,6 +79,8 @@ type (
 func (*X) TypeName() string { return "XXX" }
 func (*W) Format() string   { return "wallet" }
 func (*W) Type() string     { return "string" }
+func (ns) Nullable() bool   { return true }
+func (ni) Nullable() bool   { return false }
```
Comment (wI2L) [inline openapi/generator_test.go:83]: Need to add a test for a pointer field that implements the `Nullable` interface whos return `false` is still marked as nullable in the spec.

**Review by wI2L — CHANGES_REQUESTED** [review]

File: openapi/generator_test.go (hunk 10 above)
Comment (lukap3) [reply openapi/generator_test.go:83]: `ni` is a pointer field (https://github.com/wI2L/fizz/pull/90/files/7117e5ded046fda57a5bfed5f4d9dba39bfc7de2..2926bb1465518a3ac0752de28ab81608e0b87d5b#diff-1aeb94cb3e94f23e1ef8738104e892722175c50655af95f3c562efc596a89b08R56)
with Nullable returning `false` and it's returning true...
Or am I missing something?

File: openapi/generator_test.go (hunk 9 above)
Comment (lukap3) [reply openapi/generator_test.go:48]: a mistake 🙈

File: openapi/generator_test.go (hunk 10 above)
Comment (wI2L) [reply openapi/generator_test.go:83]: My bad, was looking at the type definition, not the field declaration.

**Comment by wI2L** [comment]
@lukap3 Thanks you for the PR.
